### PR TITLE
Document the federation remote join handshake

### PR DESCRIPTION
**Pull request by leonerd**

### CODE REVIEW

**Comment by leonerd** [comment]
@erikjohnston - this is about all I can do for now, without some input from you on these remaining TODO issues.

**Comment by leonerd** [comment]
@erikjohnston - this is about all I can do for now, without some input from you on these remaining TODO issues.

**Comment by erikjohnston** [comment]
That mostly looks fine. A couple of notes:

- The directory server only gets involved if joining via a room alias. Servers can join rooms without using aliases, e.g. with invites.

> ... it then needs to obtain enough of the current state of the room to fill in the required fields of the ``m.room.member`` event.

I would prefer not using the term "current state" here, as that implies that the `/make_join` returns a list of state events rather than simply a proto event.

**Comment by leonerd** [comment]
@erikjohnston PTAL

File: specification/server_server_api.rst [hunk 1]
```diff
@@ -533,6 +533,161 @@ part of the path specifies the kind of query being made, and its query
...
+to obtain enough information about the room to fill in the required fields of
+the ``m.room.member`` event. It obtains this by selecting a resident from the
+candidate list, and requesting the ``make_join`` endpoint using a ``GET``
+request, specifying the room ID and the user ID of the new member who is
+attempting to join.
+
+The resident server replies to this request with a JSON-encoded object having a
+single key called ``event``; within this is an object whose fields contain some
+of the information that the joining server will need. Despite its name, this
+object is not a full object; notably it does not need to be hashed or signed by
```
Comment (richvdh) [inline specification/server_server_api.rst]: s/full object/full event/ ?

File: specification/server_server_api.rst [hunk 2]
```diff
@@ -533,6 +533,161 @@ part of the path specifies the kind of query being made, and its query
...
+``auth_events``      String   An event-reference list containing the
+                              authorization events that would allow this member
+                              to join
+``content``          Object   The event content
+``depth``            Integer  (this field must be present but is ignored; it
+                              may be 0)
+``event_id``         String   A new event ID specified by the assisting
+                              resident
+``origin``           String   The name of the assisting resident homeserver
+``origin_server_ts`` Integer  A timestamp added by the resident homeserver
```
Comment (richvdh) [inline specification/server_server_api.rst]: could we have some consitency between 'assisting resident/resident homeserver/assisting resident homeserver'?

File: specification/server_server_api.rst (hunk 1 above)
Comment (leonerd) [inline specification/server_server_api.rst]: Fixed 36af793

File: specification/server_server_api.rst (hunk 2 above)
Comment (leonerd) [inline specification/server_server_api.rst]: There is *some* consistency. Perhaps you'd like more? ;)

Done as of 923f05e

**Comment by richvdh** [comment]
I think you need Erik for your questions really, but I'd suggest that a doc with TODOs for the unknowns is better than no doc and as such +1 on the PR

**Comment by leonerd** [comment]
OK, I've deleted the TODO notes from the PR now they're all either fixed, or represented as TODOs in the spec.

@erikjohnston PTAL again then.

File: specification/server_server_api.rst [hunk 3]
```diff
@@ -533,6 +533,161 @@ part of the path specifies the kind of query being made, and its query
...
+that room, and uses it to assist in the joining process. This is the remote
+join handshake.
+
+This handshake involves the homeserver of the new member wishing to join
+(referred to here as the "joining" server), the directory server hosting the
+room alias the user is requesting to join with, and a homeserver where existing
+room members are already present (referred to as the "resident" server).
+
+In summary, the remote join handshake consists of the joining server querying
+the directory server for information about the room alias; receiving a room ID
```
Comment (illicitonion) [inline specification/server_server_api.rst]: It's worth mentioning that you can remote-join by ID no alias, if you know a join candidate list.

File: specification/server_server_api.rst [hunk 4]
```diff
@@ -533,6 +533,161 @@ part of the path specifies the kind of query being made, and its query
...
+                 send_join request ----------------------->
+                 <------------------------------- send_join response
+                 |
+  <---------- join response
+
+The first part of the handshake usually involves using the directory server to
+request the room ID and join candidates. This is covered in more detail on the
+directory server documentation, below. In the case of a new user joining a
+room as a result of a received invite, the joining user's homeserver could
+optimise this step away by picking the origin server of that invite message as
```
Comment (illicitonion) [inline specification/server_server_api.rst]: It should be noted that this server may have left the room, and *only* using this server may be buggy.

File: specification/server_server_api.rst [hunk 5]
```diff
@@ -533,6 +533,161 @@ part of the path specifies the kind of query being made, and its query
...
+The resident server replies to this request with a JSON-encoded object having a
+single key called ``event``; within this is an object whose fields contain some
+of the information that the joining server will need. Despite its name, this
+object is not a full event; notably it does not need to be hashed or signed by
+the resident homeserver. The required fields are:
+
+==================== ======== ============
+ Key                  Type     Description
+==================== ======== ============
+``type``             String   The value ``m.room.member``
```
Comment (illicitonion) [inline specification/server_server_api.rst]: We generally use lower case string in the spec (and for other types, too)

File: specification/server_server_api.rst [hunk 6]
```diff
@@ -533,6 +533,161 @@ part of the path specifies the kind of query being made, and its query
...
+``auth_events``      String   An event-reference list containing the
```
Comment (illicitonion) [inline specification/server_server_api.rst]: This isn't a string.

What *is* it, though? An array of event IDs? An array of events?

File: specification/server_server_api.rst [hunk 7]
```diff
@@ -533,6 +533,161 @@ part of the path specifies the kind of query being made, and its query
...
+                              authorization events that would allow this member
+                              to join
+``content``          Object   The event content
+``depth``            Integer  (this field must be present but is ignored; it
+                              may be 0)
+``event_id``         String   A new event ID specified by the resident
+                              homeserver
+``origin``           String   The name of the resident homeserver
+``origin_server_ts`` Integer  A timestamp added by the resident homeserver
+``prev_events``      List     An event-reference list containing the immediate
```
Comment (illicitonion) [inline specification/server_server_api.rst]: s/List/[object]/ or s/List/array/

File: specification/server_server_api.rst [hunk 8]
```diff
@@ -533,6 +533,161 @@ part of the path specifies the kind of query being made, and its query
...
+fields.
+
+To complete the join handshake, the joining server must now submit this new
+event to an resident homeserver, by using the ``send_join`` endpoint. This is
+invoked using the room ID and the event ID of the new member event.
+
+The resident homeserver then accepts this event into the room's event graph,
+and responds to the joining server with the full set of state for the newly-
+joined room. This is returned as a two-element list, whose first element is the
+integer 200, and whose second element contains the following keys:
```
Comment (illicitonion) [inline specification/server_server_api.rst]: s/element contains/element is an object which contains/

**Comment by illicitonion** [comment]
Why no swagger?

File: specification/server_server_api.rst (hunk 3 above)
Comment (leonerd) [inline specification/server_server_api.rst]: That's mentioned in two places already, both after this.

File: specification/server_server_api.rst (hunk 4 above)
Comment (leonerd) [inline specification/server_server_api.rst]: Done 122c082

File: specification/server_server_api.rst (hunk 5 above)
Comment (leonerd) [inline specification/server_server_api.rst]: Really? I was copying the prevailing style in other parts of this document, and all of those others are capitalised.

File: specification/server_server_api.rst (hunk 6 above)
Comment (leonerd) [inline specification/server_server_api.rst]: It's a list of two-element [string, string] lists. I don't know how to notate that here.

Changed to `List` in 6cbfba7

File: specification/server_server_api.rst (hunk 8 above)
Comment (leonerd) [inline specification/server_server_api.rst]: Done 233e848

**Comment by leonerd** [comment]
I don't know what swagger is, or what part it would play here. Can you expand?

**Comment by leonerd** [comment]
Having now discovered what swagger actually is, I've now worked out that there is nothing about server-server federation in there at all. At @erikjohnston's suggestion I'm going to commit on what we have, and we can add more later.

**Pull request merged**